File: WuchangLiveWallpaper/app/src/main/java/com/halot455/wuchangwallpaper/WuchangWallpaperService.java
package com.halot455.wuchangwallpaper;

import android.graphics.Canvas;
import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;
import android.service.wallpaper.WallpaperService;
import android.view.SurfaceHolder;

public class WuchangWallpaperService extends WallpaperService {
    @Override public Engine onCreateEngine() { return new EngineImpl(); }

    private final class EngineImpl extends Engine {
        private final Handler h = new Handler(Looper.getMainLooper());
        private final HeroRenderer renderer = new HeroRenderer(WuchangWallpaperService.this);
        private boolean visible;
        private float angle;
        private float xOffset = .5f;
        private long last = SystemClock.uptimeMillis();

        private final Runnable tick = new Runnable() {
            @Override public void run() {
                draw();
                if (visible) h.postDelayed(this, 33);
            }
        };

        @Override public void onVisibilityChanged(boolean v) {
            visible = v;
            h.removeCallbacks(tick);
            if (v) { last = SystemClock.uptimeMillis(); tick.run(); }
        }

        @Override public void onSurfaceChanged(SurfaceHolder holder, int format, int width, int height) {
            super.onSurfaceChanged(holder, format, width, height);
            draw();
        }

        @Override public void onSurfaceDestroyed(SurfaceHolder holder) {
            visible = false;
            h.removeCallbacks(tick);
            super.onSurfaceDestroyed(holder);
        }

        @Override public void onOffsetsChanged(float xOffset, float yOffset, float xStep, float yStep, int xPixels, int yPixels) {
            this.xOffset = xOffset;
            draw();
        }

        private void draw() {
            SurfaceHolder holder = getSurfaceHolder();
            Canvas c = null;
            try {
                c = holder.lockCanvas();
                if (c == null) return;
                long now = SystemClock.uptimeMillis();
                angle = (float)Math.sin(now * 0.00048f) * 25f;
                float displayAngle = angle + (xOffset - .5f) * 12f;
                String character = getSharedPreferences("wuchang", MODE_PRIVATE).getString("character", "white");
                renderer.draw(c, c.getWidth(), c.getHeight(), displayAngle, now, character, xOffset);
            } finally {
                if (c != null) holder.unlockCanvasAndPost(c);
            }
        }
    }
}

File: WuchangLiveWallpaperKotlin/app/src/main/java/com/halot455/wuchangwallpaper/WuchangWallpaperService.kt
package com.halot455.wuchangwallpaper

import android.graphics.Canvas
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.service.wallpaper.WallpaperService
import android.view.SurfaceHolder
import kotlin.math.sin

class WuchangWallpaperService : WallpaperService() {
    override fun onCreateEngine(): Engine = LiveEngine()

    inner class LiveEngine : Engine() {
        private val handler = Handler(Looper.getMainLooper())
        private val renderer = HeroRenderer(this@WuchangWallpaperService)
        private var visible = false
        private var pageOffset = 0.5f

        private val frame = object : Runnable {
            override fun run() {
                renderFrame()
                if (visible) handler.postDelayed(this, 33L)
            }
        }

        override fun onVisibilityChanged(visibleNow: Boolean) {
            visible = visibleNow
            handler.removeCallbacks(frame)
            if (visible) frame.run()
        }

        override fun onSurfaceChanged(
            holder: SurfaceHolder,
            format: Int,
            width: Int,
            height: Int
        ) {
            super.onSurfaceChanged(holder, format, width, height)
            renderFrame()
        }

        override fun onSurfaceDestroyed(holder: SurfaceHolder) {
            visible = false
            handler.removeCallbacks(frame)
            super.onSurfaceDestroyed(holder)
        }

        override fun onOffsetsChanged(
            xOffset: Float,
            yOffset: Float,
            xOffsetStep: Float,
            yOffsetStep: Float,
            xPixelOffset: Int,
            yPixelOffset: Int
        ) {
            pageOffset = xOffset
            if (visible) renderFrame()
        }

        private fun renderFrame() {
            val holder = surfaceHolder
            var canvas: Canvas? = null
            try {
                canvas = holder.lockCanvas() ?: return
                val now = SystemClock.uptimeMillis()
                val character = getSharedPreferences("wuchang", MODE_PRIVATE)
                    .getString("character", "white") ?: "white"
                val idle = sin(now * 0.00048).toFloat() * 24f
                val parallax = (pageOffset - 0.5f) * 10f
                renderer.draw(
                    canvas,
                    canvas.width,
                    canvas.height,
                    idle + parallax,
                    now,
                    character,
                    pageOffset
                )
            } catch (_: Throwable) {
            } finally {
                canvas?.let {
                    try {
                        holder.unlockCanvasAndPost(it)
                    } catch (_: Throwable) {
                    }
                }
            }
        }
    }
}
